# P-GRe v.0

WORK IN PROGRESS ! Check later :-)

Contacts and authors
====================
Sébastien Cabanac, PhD student, sebastien.cabanac@univ-tlse3.fr

Christophe Dunand, professor at Université Paul Sabatier Toulouse 3, christophe.dunand@univ-tlse3.fr

Catherine Mathé, associate professor at Université Paul Sabatier Toulouse 3, catherine.mathe-dehais@univ-tlse3.fr

All members of the Laboratoire de Recherche en Sciences Végétales, Université de Toulouse, CNRS, Université Paul Sabatier Toulouse 3, Toulouse INP, Auzeville-Tolosane, France.

Funding and acknowledgements
============================
The authors are thankful to the Paul Sabatier-Toulouse 3 University and to the Centre National de la Recherche Scientifique (CNRS) for granting their work. Sébastien Cabanac is the recipient of a fellowship from the “École Universitaire de Recherche (EUR)” TULIP-GS (ANR-18-EURE-0019). This study is set within the framework of the “Laboratoires d’Excellences (LABEX)” TULIP (ANR-10-LABX-41).

Contents
========

- [Contacts and authors](#contacts-and-authors)
- [Funding and acknowledgements](#funding-and-acknowledgements)
- [About P-GRe](#about-p-gre)
- [Requirements and input](#requirements-and-input)
- [How P-GRe works](#how-p-gre-works)
- [Installation](#installation)
- [Running P-GRe](#running-p-gre)
- [Common problems](#common-problems)
- [Citing P-GRe and associated software](#citing-p-gre-and-associated-software)
- [Licence](#licence)
- [References](#references)

About P-GRe
===========
Pseudogenes are genomic sequences with homology to functional genes but that harbor deleterious mutations, such as loss of the start codon, loss of coding sequence, gain of stop or frame-shifts. No longer coding for a functional protein, pseudogenes are rarely transcribed and are often described as having no function. It is now known that part of the pseudogenes are transcribed, this part representing for example 15% of all the pseudogenes in mice<sup name="a1">[R1](#f1)</sup>. It has also been shown that these transcripts, originating from pseudogenes, could form duplexes with the mRNAs of homologous functional genes and thus participate in post-transcriptional regulation through the RNAi pathway <sup name="a2">[R2, ](#f2)</sup><sup name="a3">[R3, ](#f3)</sup><sup name="a4">[R4](#f4)</sup>. In the other hand, the exhaustive prediction of pseudogenes allow a better understanding of dynamic of gene evolution in multigenenic families often subjected to duplication and pseudogenisation events.

The goal of PseudoGene REtriever (P-GRe [/pɛɡʁ/]) is to find the position of pseudogenes on a genome, as well as to infer their structures in pseudo-exons and pseudo-introns. P-GRe aims to be more user-friendly, with a limited number of dependencies, ease of use and total automaticity, while producing qualitative results and having greater sensitivity than other software with the same goal.

Requirements and input
======================



How P-GRe works
===============


Installation
============



Running P-GRe
=============

Common problems
===============

To be added.

Citing P-GRe and associated software
====================================

To be added.

Licence
=======
All source code contained in the `scripts` folder and the `P-GRe_pipeline.sh` file are under the [Artistic Licence 1.0](https://opensource.org/license/artistic-1-0/).

References
==========
<b id="f1">[R1]</b> C. Sisu *et al*. Transcriptional activity and strain-specific history of mouse pseudogenes. *Nat Commun*. Vol. 11, 2020.[↩](#a1)

<b id="f2">[R2]</b> O. H. Tam *et al*. Pseudogene-derived small interfering RNAs regulate gene expression in mouse oocytes. *Nature*. Vol. 453, pp. :534-538, 2008.[↩](#a2)

<b id="f3">[R3]</b> T. Watanabe *et al*. Endogenous siRNAs from naturally formed dsRNAs regulate transcripts in mouse oocytes. *Nature*. Vol. 453, pp. :539-543, 2008.[↩](#a3)

<b id="f4">[R4]</b> P. Guo *et al*. Engineering RNA for Targeted siRNA Delivery and Medical Application. *Adv Drug Deliv Rev*. Vol. 62, pp. :650-666, 2010.[↩](#a4)
